hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/ModifyTableProcedure.java

```diff
@@ -506,14 +506,17 @@ private void removeReplicaColumnsIfNeeded(MasterProcedureEnv env) throws IOExcep
   private void assignNewReplicasIfNeeded(MasterProcedureEnv env) throws IOException {
     final int oldReplicaCount = unmodifiedTableDescriptor.getRegionReplication();
     final int newReplicaCount = modifiedTableDescriptor.getRegionReplication();
-    if (newReplicaCount <= oldReplicaCount) {
+    int existingReplicasCount = env.getAssignmentManager().getRegionStates()
+      .getRegionsOfTable(modifiedTableDescriptor.getTableName()).size();
+    if (newReplicaCount <= Math.min(oldReplicaCount, existingReplicasCount)) {
```

Comment:
Why do we need to consider the ZK info here? Shouldn't the desc diff be enough?

Reply:
In the problematic case we have the same (3) replica count in the Table Descriptor and the problem is that only 1 of them exists in ZK. After this we will assign the other 2 as well.

In my previous implementation I changed the replica count in the TableDescriptor so it was not needed.

Reply:
After we run the createAssignProcedures for the 2 replicas, ZK info will be correct too.
In the problematic case the oldReplicaCount and newReplicaCount are both 3(the same) so the replica assignment wouldn't happen.

Reply:
Got it, yeah it seems we need to check for differences between the two descs and zk.